### PR TITLE
feat(*): convert to typescript

### PR DESCRIPTION
**Pull request by straker**
This one was a bit tricky with all the private property lookups in axe-core and React.

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Code is reviewed for security

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,394 @@
+/* global Promise */
+import axeCore = require('axe-core');
+import rIC = require('requestidlecallback');
+import after = require('./after');
+
+// @reference https://www.npmjs.com/package/@types/requestidlecallback
+// (it's declared as type=dom so had to copy/paste here)
```
Comment (stephenmathieson) [inline index.ts]: What does "it's declared as type=dom" mean? Can't we just add `"dom"` to `lib: []`?

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+        };
+        none: {
+          failureMessage: (args: string[]) => string;
+        };
+      };
+    };
+  };
+}
+
+const requestIdleCallback: requestIdleCallback = rIC.request;
```
Comment (stephenmathieson) [inline index.ts]: Having a variable with the same name as a type is confusing. I'm very surprised ESLint didn't yell about this.

Please consider renaming the type.

File: package.json [hunk 3]
```diff
@@ -40,10 +41,13 @@
   "author": "Dylan Barrell (dylan@barrell.com)",
   "license": "MPL-2.0",
   "dependencies": {
+    "@types/react": "^16.9.20",
```
Comment (stephenmathieson) [inline package.json]: This should be a dev dep.

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+const critical = 'color:red;font-weight:bold;';
+const serious = 'color:red;font-weight:normal;';
+const moderate = 'color:orange;font-weight:bold;';
+const minor = 'color:orange;font-weight:normal;';
+const defaultReset = 'font-color:black;font-weight:normal;';
+
+let idleId: number | undefined;
+let timeout: number;
+let context: axeCore.ElementContext;
+let _createElement: Function;
```
Comment (stephenmathieson) [inline index.ts]: `typeof React.createElement` (or something, I don't remember how to do this) might be better here. Using `Function` is pretty ambiguous.

File: index.ts [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+              const key: string = node.target.toString() + result.id;
+              const retVal = !cache[key];
+              cache[key] = key;
+              return retVal;
+            });
+            return !!result.nodes.length;
+          });
+
+          if (results.violations.length) {
+            console.group('%cNew aXe issues', serious);
```
Comment (scurker) [inline index.ts]: Unrelated to your change, but we probably want the correct _axe_ branding here

File: index.ts [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+ * Check the node for violations.
+ * @param {Component} component
+ */
+function checkNode(component: React.Component): void {
+  let node: Node;
+
+  try {
+    node = ReactDOM.findDOMNode(component);
+  } catch (e) {
+    console.group('%caXe error: could not check node', critical);
```
Comment (scurker) [inline index.ts]: Branding: axe

**Comment by dgrant85** [comment]
Has this been released?

Version pulled from npm does not include type declarations and is listed on npmjs.com as 3.4.1, whereas develop is currently at 3.4.0?

**Comment by straker** [comment]
It does not seem types was released to npm. As for the version discrepancy, seems we never pulled master back into develop after the release.

**Pull request merged**